prov/hook/hook_hmem/src/hook_hmem.c

```diff
@@ -287,32 +291,39 @@ static int hook_hmem_track_atomic(struct hook_ep *ep, const struct fi_ioc *ioc,
 	if (res_count) {
 		(*hmem_ctx)->res_desc = calloc(res_count,
 					     sizeof(**(*hmem_ctx)->res_desc));
-		if (!(*hmem_ctx)->res_desc)
-			goto err1;
+		if (!(*hmem_ctx)->res_desc) {
+			ret = -FI_ENOMEM;
+			goto err4;
+		}
 
 		ofi_ioc_to_iov(res_ioc, res_iov, res_count, dt_size);
 		ret = hook_hmem_cache_mr_iov(domain, res_iov, res_desc,
 					res_count, (*hmem_ctx)->res_desc);
 		if (ret) {
 			free((*hmem_ctx)->comp_desc);
-			goto err1;
+			goto err4;
 		}
 		(*hmem_ctx)->res_count = res_count;
 	} else {
 		(*hmem_ctx)->res_count = 0;
 		(*hmem_ctx)->res_desc = NULL;
 	}
 
-	goto out;
-err1:
-	hook_hmem_uncache_mr_iov(domain, comp_iov, comp_count);
-	free((*hmem_ctx)->comp_desc);
-err2:
-	hook_hmem_uncache_mr_iov(domain, iov, count);
+	ofi_mutex_unlock(&domain->lock);
+	return FI_SUCCESS;
+
+err4:
+	if (comp_count) {
+		hook_hmem_uncache_mr_iov(domain, comp_iov, comp_count);
+		free((*hmem_ctx)->comp_desc);
+	}
 err3:
+	if (count)
+		hook_hmem_uncache_mr_iov(domain, iov, count);
+err2:
 	ofi_buf_free(*hmem_ctx);
-out:
-	ofi_mutex_lock(&domain->lock);
+err1:
+	ofi_mutex_unlock(&domain->lock);
```

Comment:
I was wondering how no one has tripped on this already when I saw this particular defect in the report.